### ISSUE TITLE
Better to depend on non-cdt versions?

### ISSUE THREAD

**Issue by hmaarrfk** [conda-forge/gtk3-feedstock#7]
Issue:
It seems that with the latest update, users must remember to install `libXi` in their `yum_requirements.txt` otherwise the post install scripts don't work.

Is it better to try and use the conda-forge X11 libraries so that these dependencies are managed automatically?
```
 ERROR conda.core.link:_execute(700): An error occurred while installing package 'conda-forge::gtk3-3.24.18-h45fd312_0'.
Rolling back transaction: ...working... done
LinkError: post-link script failed for package conda-forge::gtk3-3.24.18-h45fd312_0
location of failed script: /opt/conda/envs/owl_tester/bin/.gtk3-post-link.sh
==> script messages <==
<None>
==> script output <==
stdout: 
stderr: gtk-update-icon-cache: Cache file created successfully.
/opt/conda/envs/owl_tester/bin/gtk-query-immodules-3.0: error while loading shared libraries: libXi.so.6: cannot open shared object file: No such file or directory
return code: 127
()
bash: cannot set terminal process group (-1): Inappropriate ioctl for device
```
```
 gtk3-3.24.18         | 19.6 MB   | ########## | 100% 
```

**Comment by tschoonj** [comment]
I think you will need to add `libXi` and friends as `cdt` build-time dependencies to your recipe, like they do in [`gtkmm-feedstock`](https://github.com/conda-forge/gtkmm-feedstock/blob/master/recipe/meta.yaml). Depending on what your tests do, you will need to add these packages to `yum_requirements.txt` as well.

**Comment by hmaarrfk** [comment]
Right, for me since I didn't link to the X11 libraries myself, all I needed was the `yum_requirements.txt` file.

I just don't know if this is a sustainable approach as more packages start to depend on gtk3.

I'm wondering if it is better to depend on conda-forge's stack.

I guess maybe not all of the packages you need are available on conda-forge

A quick search couldn't find the x11xi package.

**Comment by stuarteberg** [comment]
FWIW, there was a related discussion in the staged-recipes repo when the gtk3 recipe was first contributed:

https://github.com/conda-forge/staged-recipes/pull/10979#issuecomment-595819565

**Comment by ryanvolz** [comment]
Absent guidance from conda-forge as a whole (https://github.com/conda-forge/cfep/pull/7, maybe a good time for someone to resurrect that discussion?), I've also been feeling the pain of making recipes with either `gtk3` or `qt` and then having to figure out what CDTs or `yum_requirements.txt` I need to add to get it to work. Pain is maybe too strong, because it amounts to copying what I've done for previous recipes, but remembering to do it is annoying. So I agree that this is definitely a problem that would be good to get ahead of before too many people adopt `gtk3`.

A bit of history: @pkgw created the `gtk3` package that I adopted and submitted to conda-forge, and his version uses the `xorg-` packages. I switched to x11 CDTs for submission because it put `gtk3` on a level playing field with the `qt` package which also uses CDTs. Some of the `gtk3` dependencies require the `xorg-` packages (namely `cairo`), so what actually happens right now is that a user will get some `xorg-` packages and anything that's missing will use the system library. So certainly I don't think that the current situation is perfect, and it is definitely possible to switch to using the `xorg-` packages.

So I see two ways to handle this:

1. Document the specific CDTs and `yum_requirements.txt` that might be needed by dependent packages. The maximal set probably can't change, and it might be possible that packages can use less than that. It's a bit different from the way packagers might expect a conda recipe to work, but it's not unheard of (`qt` has the same problem). I don't know where this might go to get good visibility, but that's probably as solvable problem.
2. Switch to using the `xorg-` packages. I would go back to the pkgw recipe, see what was needed there, and plug those in. Then the normal package dependencies will take care of things for downstream packages, with no surprises. What I don't know is if there are any technical or performance limitations by going this route.

Thoughts?

**Comment by hmaarrfk** [comment]
Thanks all for the history guys. I'm not sure what the best approach is so I'm going to stay silent.

It was just suprizing for me to see see some of my builds fail this weekend as they were depending on gtk3.
I found a workaround that worked for me.

Curious to learn what you all will decide.

**Comment by stuarteberg** [comment]
>I've also been feeling the pain of making recipes with either gtk3 or qt and then having to figure out what CDTs or yum_requirements.txt I need to add to get it to work.

Me too, recently.  One thing that could make it less painful is emitting an error message telling you exactly which libraries you're missing.  That's similar to the documentation you propose, but it would be harder to miss.

I'm not sure exactly how to implement it.  Maybe a `post-link.sh` script that runs `ldd` on all `gtk3` libraries to figure out what the user is missing?  All users would benefit from such a message.  An additional hint about `yum_requirements.txt` could help conda-forge recipe authors.

**Comment by pkgw** [comment]
I'll just chime in to say that I'd be more than happy if someone could come in and supersede conda-forge/cfep#7. The conda-forge project has gotten pretty far in the current situation, but fundamentally our story about X11 is muddled and it would be good to resolve that.

I put the `xorg-*` stack together before the whole CDT concept existed, and while the CDT scheme has its rough edges (as documented here) there are good reasons for its existence and I don't see it going away any time soon. So I guess I'd recommend putting the effort into thinking about how to smooth out the CDT side of things rather than adding deps on the `xorg-*` stack.

**Comment by ryanvolz** [comment]
I'm going to re-start this discussion in light of what we've done with `gtk4` (not use CDTs and instead depend on the conda-forge xorg packages) and the continuing annoyances like #16.

I'm in favor of switching out these CDT dependencies at this point. I think the maintenance burden on everyone, particularly downstream packages, has proven to be large enough to warrant depending on the conda-forge xorg packages despite any potential drawbacks. Does anybody else have an opinion on this?

**Comment by ryanvolz** [comment]
Implemented in #59.
